Common/Util/RateLimit/ThreadSleepStrategy.cs

```diff
@@ -0,0 +1,60 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System.Threading;
+
+namespace QuantConnect.Util.RateLimit
+{
+    /// <summary>
+    /// Provides a CPU non-intensive means of waiting for more tokens to be available in <see cref="ITokenBucket"/>.
+    /// This strategy should be the most commonly used as it either sleeps or yields the currently executing thread,
+    /// allowing for other threads to execute while the current thread is blocked and waiting for new tokens to become
+    /// available in the bucket for consumption.
+    /// </summary>
+    public class ThreadSleepStrategy : ISleepStrategy
```

Comment:
If we are not going to be using the blocking `Consume` implementation I'm not sure I see the value on these implementations/interfaces, I'd be more inclined to simplify if possible

Reply:
In general, I think it's _very_ beneficial to design concepts and solutions in isolation from the current task. My general design philosophy works by first identifying the problem(s) that need to be solved. In this case, my design called for a leaky bucket implementation, so the first major commit in this PR is indeed, a full featured leaky bucket algorithm. After designing the component _at the core_ of this entire PR, I then proceeded to implement the specifics of the problem at hand.

Over the years I've learned that there is certainly balance between _just enough_ and *way* too much. In the case of these implementations, they're ported from the referenced github library, so they don't cost us anything, but provide solid abstractions and implementations to act as fundamental building blocks for future solutions. While I don't think everything should be componentized, I do find that development continues to get easier and easier if you follow this kind of approach due to the quality of the building blocks available to you.

Reply:
Wondering which is the `referenced github library`?

Reply:
I left a note in the source [here](https://github.com/QuantConnect/Lean/pull/3743/files#diff-dc9ad5005cc9b09a90ed99a877a77eb2R26). It's based on this [`TokenBucket`](https://github.com/mxplusb/TokenBucket) dotnet core repository.